geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java

```diff
@@ -127,50 +127,139 @@ private int getArrayIndex(int listIndex) {
     return elementList.size();
   }
 
   /**
    * @param region the region this instance is stored in
-   * @param key the name of the set to add to
+   * @param key the name of the list to add to
    * @return the element actually popped
    */
   public byte[] lpop(Region<RedisKey, RedisData> region, RedisKey key) {
-    byte[] popped = elementRemove(0);
-    RemoveElementsByIndex removed = new RemoveElementsByIndex();
-    removed.add(0);
+    byte newVersion;
+    byte[] popped;
+    RemoveElementsByIndex removed;
+    synchronized (this) {
+      newVersion = incrementAndGetVersion();
+      popped = removeFirstElement();
+      removed = new RemoveElementsByIndex(newVersion);
+      removed.add(0);
+    }
     storeChanges(region, key, removed);
     return popped;
   }
 
+  public synchronized byte[] removeFirstElement() {
+    return elementList.removeFirst();
+  }
+
+  public synchronized byte[] removeLastElement() {
+    return elementList.removeLast();
+  }
+
   /**
-   * @return the number of elements in the list
+   * @param elementsToAdd elements to add to this list; NOTE this list may be modified by this call
+   * @param region the region this instance is stored in
+   * @param key the name of the list to add to
+   * @param onlyIfExists if true then the elements should only be added if the key already exists
+   *        and holds a list, otherwise no operation is performed.
+   * @return the length of the list after the operation
    */
-  public int llen() {
+  public long lpush(List<byte[]> elementsToAdd, Region<RedisKey, RedisData> region, RedisKey key,
+      final boolean onlyIfExists) {
+    byte newVersion;
+    AddByteArrays addByteArrays;
+    synchronized (this) {
+      newVersion = incrementAndGetVersion();
+      elementsPush(elementsToAdd);
+      addByteArrays = new AddByteArrays(newVersion, elementsToAdd);
+    }
+    storeChanges(region, key, addByteArrays);
     return elementList.size();
   }
 
+  /**
+   * @param elementsToAdd elements to add to this list; NOTE this list may be modified by this call
+   * @param region the region this instance is stored in
+   * @param key the name of the list to add to
+   * @return the number of elements actually added
+   */
+  public long lpush(List<byte[]> elementsToAdd, Region<RedisKey, RedisData> region, RedisKey key) {
+    byte newVersion;
+    AddByteArrays addByteArrays;
+    synchronized (this) {
+      newVersion = incrementAndGetVersion();
+      elementsPush(elementsToAdd);
+      addByteArrays = new AddByteArrays(newVersion, elementsToAdd);
+    }
+    storeChanges(region, key, addByteArrays);
+    return elementList.size();
+  }
+
+  public byte[] ltrim(long start, long end, Region<RedisKey, RedisData> region,
+      RedisKey key) {
+    byte newVersion;
+    int length = elementList.size();
+    int boundedStart = getBoundedStartIndex(start, length);
+    int boundedEnd = getBoundedEndIndex(end, length);
+    List<Integer> removed = new ArrayList<>();
```

Comment:
It does not seem performant to build up an ArrayList of every index we are going to remove. All you need for this DeltaInfo is two numbers, start and end. So just add an RetainElementsByIndexRange DeltaInfo.
You can make sure they are actual "bounded" indexes.
Then instead of elementsRemove having to iterate over the list for very item you are removing, you can just implement elementsRetainByIndexRange that takes two ints. You should be able to implement that method by using an Iterator that only traverse the linked list once.

Reply:
Got a different scheme now, just sending the start and end indexes over the wire (one of them negative if the whole list is gone).